kaldb/src/main/java/com/slack/kaldb/chunkManager/ChunkManagerBase.java

```diff
@@ -107,72 +84,88 @@ public SearchResult<T> query(SearchQuery query, Duration queryTimeout) {
-                                // invalid queries are received
-                                LOG.warn("Chunk Query Exception: {}", err.getMessage());
-                                LOG.debug("Chunk Query Exception", err);
-                                // We catch IllegalArgumentException ( and any other exception that
-                                // represents a parse failure ) and instead of returning an empty
-                                // result we throw back an error to the user
-                                if (err instanceof IllegalArgumentException) {
-                                  throw err;
-                                }
-                                return errorResult;
-                              }
-                            })))
-            .peek(
-                (future) ->
-                    queryCancellationService.schedule(
-                        () -> future.cancel(true), queryTimeout.toMillis(), TimeUnit.MILLISECONDS))
-            .collect(Collectors.toList());
-
-    Future<List<SearchResult<T>>> searchResultFuture = Futures.successfulAsList(queries);
     try {
-      List<SearchResult<T>> searchResults =
-          searchResultFuture.get(queryTimeout.toMillis(), TimeUnit.MILLISECONDS);
-
-      // check if all results are null, and if so return an error to the user
-      if (searchResults.size() > 0 && searchResults.stream().allMatch(Objects::isNull)) {
+      try (var scope = new StructuredTaskScope<SearchResult<T>>()) {
+        List<StructuredTaskScope.Subtask<SearchResult<T>>> chunkSubtasks =
+            chunksMatchingQuery.stream()
+                .map(
+                    (chunk) ->
+                        scope.fork(
+                            currentTraceContext.wrap(
+                                () -> {
+                                  ScopedSpan span =
+                                      Tracing.currentTracer()
+                                          .startScopedSpan("ChunkManagerBase.chunkQuery");
+                                  span.tag("chunkId", chunk.id());
+                                  span.tag("chunkSnapshotPath", chunk.info().getSnapshotPath());
+                                  concurrentQueries.acquire();
+                                  try {
+                                    return chunk.query(query);
+                                  } finally {
+                                    concurrentQueries.release();
+                                    span.finish();
+                                  }
+                                })))
+                .toList();
         try {
-          Futures.allAsList(queries).get(0, TimeUnit.SECONDS);
-        } catch (Exception e) {
-          throw new IllegalArgumentException(e);
+          scope.joinUntil(Instant.now().plusSeconds(queryTimeout.toSeconds()));
+        } catch (TimeoutException timeoutException) {
+          scope.shutdown();
+          scope.join();
```

Comment:
maybe we can add some docs as to why we call `join` here? This definitely is something I hadn't realized in the past :)